File: HabitTracker.Core/Models/Habit.cs
using System.ComponentModel.DataAnnotations;

namespace HabitTracker.Core.Models;

public class Habit
{
    public int Id { get; set; }

    [Required]
    [MaxLength(100)]
    public string Name { get; set; } = string.Empty;

    [MaxLength(500)]
    public string Description { get; set; } = string.Empty;

    [Required]
    public FrequencyType Frequency { get; set; } = FrequencyType.Daily;

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public bool IsArchived { get; set; }

    public bool IsCompleted { get; set; }

    public int? CategoryId { get; set; }
    public Category? Category { get; set; }

    public int? TargetDays { get; set; }
    public DateTime? StartDate { get; set; }
    public DateTime? EndDate { get; set; }
    public string? Reminder { get; set; }

    public ICollection<HabitProgress> Progress { get; set; } = new List<HabitProgress>();
    public string UserId { get; set; } = string.Empty;
}

public enum FrequencyType
{
    Daily,
    Weekly,
    Monthly
}

File: habit-tracker-client/src/types/habit.ts
export enum FrequencyType {
  Daily = 'Daily',
  Weekly = 'Weekly',
  Monthly = 'Monthly'
}

export interface Category {
  id: number;
  name: string;
  color: string;
  icon?: string;
}

export interface Habit {
  id: number;
  name: string;
  description: string;
  frequency: FrequencyType;
  createdAt: string;
  isArchived: boolean;
  isCompleted: boolean;
  categoryId?: number;
  category?: Category;
  targetDays?: number;
  startDate?: string;
  endDate?: string;
  reminder?: string;
}

export interface HabitProgress {
  id: number;
  habitId: number;
  date: string;
  isCompleted: boolean;
  notes?: string;
}

export interface CreateHabitDto {
  name: string;
  description: string;
  frequency: FrequencyType;
  categoryId?: number;
  targetDays?: number;
  startDate?: string;
  endDate?: string;
  reminder?: string;
}

export interface UpdateHabitDto {
  name: string;
  description: string;
  frequency: FrequencyType;
  isArchived: boolean;
  isCompleted: boolean;
  categoryId?: number;
  targetDays?: number;
  startDate?: string;
  endDate?: string;
  reminder?: string;
}

export interface HabitStatistics {
  habitId: number;
  habitName: string;
  currentStreak: number;
  longestStreak: number;
  completionRate: number;
  totalCompletions: number;
  daysTracked: number;
  completionsByDay: Record<string, number>;
}

export interface CategoryStatistics {
  categoryId: number;
  categoryName: string;
  totalHabits: number;
  activeHabits: number;
  averageCompletionRate: number;
  habitsByFrequency: Record<string, number>;
}

export interface UserStatistics {
  totalHabits: number;
  activeHabits: number;
  totalCategories: number;
  overallCompletionRate: number;
  completionRateByCategory: Record<string, number>;
  topPerformingHabits: HabitStatistics[];
  needsAttentionHabits: HabitStatistics[];
}
